Add stream socket so_linger context option

bukka opened the pull request:
This adds so_linger stream socket context options that is used to set the lingering time in seconds. If the value is lower or equal to 0, then the lingering is disabled.

bukka committed:
Add stream socket so_linger context option

This adds so_linger stream socket context options that is used to set
the lingering time in seconds. If the value is lower or equal to 0,
then the lingering is disabled.

bukka committed:
Use SO_LINGER_SEC if available (MacOS)

bukka committed:
Do not relay on exact l_onoff value (FreeBSD uses 128 instead of 1)

devnexen commented on main/network.c:
```diff
@@ -541,6 +541,19 @@ php_socket_t php_network_bind_socket_to_local_addr_ex(const char *host, unsigned
 		if (sockvals != NULL) {
+#ifdef SO_LINGER
+			if (sockvals->mask & PHP_SOCKVAL_SO_LINGER) {
+				struct linger linger_val = {
```
what can be done eventualy here, is something like this:

```c
unsigned short secs = sockvals->linger > USHRT_MAX ? USHRT_MAX : (unsigned short)sockvals->linger;
struct linger linger_val = {
    .l_onoff = (sockvals->linger > 0),
	.l_linger = sockvals->linger > 0 ? secs : 0,
};
```

devnexen commented on ext/standard/tests/network/so_linger.phpt:
```diff
@@ -0,0 +1,77 @@
+    die('Unable to accept connection');
+}
+
+$so_linger = PHP_OS_FAMILY === 'Darwin' ? SO_LINGER_SEC : SO_LINGER;
```
could be instead

```
$so_linger = defined('SO_LINGER_SEC') : SO_LINGER_SEC : SO_LINGER;
```

devnexen committed:
Clamp so_linger context option instead of truncating it

65536 became l_onoff=1, l_linger=0, an abortive close. Also fix the test's
SO_LINGER_SEC detection and drop its unreachable SKIPIF.

devnexen committed:
Add NEWS/UPGRADING entries for so_linger

bukka committed:
Merge branch 'master' into stream_so_linger
